test/e2e/framework/pod/resize.go

```diff
@@ -0,0 +1,470 @@
+		Args:         []string{"-c", cmd},
+		Resources:    res,
+		ResizePolicy: resizePol,
+	}
+
+	tcStatus := v1.ContainerStatus{
+		Name:               tcInfo.Name,
+		AllocatedResources: alloc,
+	}
+	return tc, tcStatus
+}
+
+func MakePodWithResizableContainers(ns, name, timeStamp string, tcInfo []ResizableContainerInfo) *v1.Pod {
+	var testContainers []v1.Container
+
+	for _, ci := range tcInfo {
+		tc, _ := makeResizableContainer(ci)
+		testContainers = append(testContainers, tc)
+	}
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+			Labels: map[string]string{
+				"time": timeStamp,
+			},
+		},
+		Spec: v1.PodSpec{
+			OS:            &v1.PodOS{Name: v1.Linux},
+			Containers:    testContainers,
+			RestartPolicy: v1.RestartPolicyOnFailure,
+		},
+	}
+	return pod
+}
+
+func VerifyPodResizePolicy(gotPod *v1.Pod, wantCtrs []ResizableContainerInfo) {
+	ginkgo.GinkgoHelper()
+	for i, wantCtr := range wantCtrs {
+		gotCtr := &gotPod.Spec.Containers[i]
+		ctr, _ := makeResizableContainer(wantCtr)
+		gomega.Expect(gotCtr.Name).To(gomega.Equal(ctr.Name))
+		gomega.Expect(gotCtr.ResizePolicy).To(gomega.Equal(ctr.ResizePolicy))
+	}
+}
+
+func VerifyPodResources(gotPod *v1.Pod, wantCtrs []ResizableContainerInfo) {
+	ginkgo.GinkgoHelper()
+	for i, wantCtr := range wantCtrs {
+		gotCtr := &gotPod.Spec.Containers[i]
+		ctr, _ := makeResizableContainer(wantCtr)
+		gomega.Expect(gotCtr.Name).To(gomega.Equal(ctr.Name))
+		gomega.Expect(gotCtr.Resources).To(gomega.Equal(ctr.Resources))
+	}
+}
+
+func VerifyPodAllocations(gotPod *v1.Pod, wantCtrs []ResizableContainerInfo) error {
+	ginkgo.GinkgoHelper()
+	for i, wantCtr := range wantCtrs {
+		gotCtrStatus := &gotPod.Status.ContainerStatuses[i]
+		if wantCtr.Allocations == nil {
+			if wantCtr.Resources != nil {
+				alloc := &ContainerAllocations{CPUAlloc: wantCtr.Resources.CPUReq, MemAlloc: wantCtr.Resources.MemReq}
+				wantCtr.Allocations = alloc
+				defer func() {
+					wantCtr.Allocations = nil
+				}()
+			}
+		}
+
+		_, ctrStatus := makeResizableContainer(wantCtr)
+		gomega.Expect(gotCtrStatus.Name).To(gomega.Equal(ctrStatus.Name))
+		if !cmp.Equal(gotCtrStatus.AllocatedResources, ctrStatus.AllocatedResources) {
+			return fmt.Errorf("failed to verify Pod allocations, allocated resources not equal to expected")
+		}
+	}
+	return nil
+}
+
+func VerifyPodStatusResources(gotPod *v1.Pod, wantCtrs []ResizableContainerInfo) {
+	ginkgo.GinkgoHelper()
+	for i, wantCtr := range wantCtrs {
+		gotCtrStatus := &gotPod.Status.ContainerStatuses[i]
+		ctr, _ := makeResizableContainer(wantCtr)
+		gomega.Expect(gotCtrStatus.Name).To(gomega.Equal(ctr.Name))
+		gomega.Expect(ctr.Resources).To(gomega.Equal(*gotCtrStatus.Resources))
+	}
+}
+
+func isPodOnCgroupv2Node(f *framework.Framework, pod *v1.Pod) bool {
+	// Determine if pod is running on cgroupv2 or cgroupv1 node
+	//TODO(vinaykul,InPlacePodVerticalScaling): Is there a better way to determine this?
```

Comment:
Maybe use the same method as libcontainer? https://github.com/kubernetes/kubernetes/blob/2049360379bcc5d6467769cef112e6e492d3d2f0/vendor/github.com/opencontainers/runc/libcontainer/cgroups/utils.go#L37

Reply:
It looks like the NPD test also branches on cgroups v2. Would be good to consolidate these helpers: https://github.com/kubernetes/kubernetes/blob/2049360379bcc5d6467769cef112e6e492d3d2f0/test/e2e/node/node_problem_detector.go#L369

Reply:
Thanks, I noticed the same. But there are 2 isuses:

1. libcontainer method does not work for cluster e2e tests because it will return the result based on the test binary's location. In case of cluster e2e tests, the test binary runs on prow job whereas in case of node e2e tests, the test binary runs on the node.

2. NPD test utility does not work for node e2e tests because it already SSH into the node.

Reply:
SSH doesn't work for the node e2e tests? I'm surprised. Maybe there's an alternative?

You can just add a TODO for now to de-duplicate them.

Reply:
Sorry, I did not explain clearly.

Node e2e tests work by copying the `e2e_node.test` binary to the node remotely using `scp`, and then executing it over `ssh`.  If we were to utilize the NPD test utility in this setup, it would essentially attempt to establish an SSH connection to itself, creating a loopback scenario similar to hairpin traffic. I'm not sure if it will work.